spi/src/main/java/org/opensearch/security/spi/resources/sharing/ShareWith.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.security.spi.resources.sharing;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.opensearch.core.common.io.stream.NamedWriteable;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.xcontent.ToXContentFragment;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+
+/**
+ * This class contains information about whom a resource is shared with and at what scope.
+ * Example:
+ * "share_with": {
+ * "read_only": {
+ * "users": [],
+ * "roles": [],
+ * "backend_roles": []
+ * },
+ * "read_write": {
+ * "users": [],
+ * "roles": [],
+ * "backend_roles": []
+ * }
```

Comment:
Did we finalize the path forward w.r.t. whether we even have scopes in p0, or do we simply have something like "full_access" as a placeholder that doesn't do anything? CC @cwperks

Reply:
https://github.com/opensearch-project/security/pull/5185#issuecomment-2744341082